### PR TITLE
Fix can not find right path when blobstore restore

### PR DESCRIPTION
**Pull request by jiaqizho**
### What problem does this PR solve?

Issue Number: ref #3594 

Problem Summary:

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Storages/Page/V3/BlobStore.cpp [hunk 1]
```diff
@@ -882,6 +911,38 @@ void BlobStore::BlobStats::restoreByEntry(const PageEntryV3 & entry)
...
+
+    String err_msg;
+    try
+    {
+        const auto & blob_id = std::stoull(ss[1]);
+        return blob_id;
+    }
+    catch (std::invalid_argument & e)
+    {
+        err_msg = e.what();
```
Comment (jiaqizho) [inline dbms/src/Storages/Page/V3/BlobStore.cpp:937]: need return `err_msg`

File: dbms/src/Storages/Page/V3/BlobStore.cpp [hunk 2]
```diff
@@ -74,6 +74,35 @@ BlobStore::BlobStore(String storage_name, const FileProviderPtr & file_provider_
...
+        {
+            const auto & [blob_id, err_msg] = BlobStats::getBlobIdFromName(blob_name);
+            if (blob_id != INVALID_BLOBFILE_ID)
+            {
+                Poco::File blob(fmt::format("{}/{}", path, blob_name));
+                delegator->addPageFileUsedSize({blob_id, 0}, blob.getSize(), path, true);
+            }
+            else
+            {
+                LOG_FMT_INFO(log, "Ignore not blob file [dir={}] [file={}][err_msg={}]", path, blob_name, err_msg);
```
Comment (JaySon-Huang) [inline dbms/src/Storages/Page/V3/BlobStore.cpp:100]:
```suggestion
                LOG_FMT_INFO(log, "Ignore not blob file [dir={}] [file={}] [err_msg={}]", path, blob_name, err_msg);
```

File: dbms/src/Storages/Page/V3/BlobStore.cpp [hunk 3]
```diff
@@ -882,6 +911,38 @@ void BlobStore::BlobStats::restoreByEntry(const PageEntryV3 & entry)
     stat->restoreSpaceMap(entry.offset, entry.size);
 }
 
+std::pair<BlobFileId, String> BlobStore::BlobStats::getBlobIdFromName(String blob_name)
+{
+    String err_msg = "no error";
```
Comment (JaySon-Huang) [inline dbms/src/Storages/Page/V3/BlobStore.cpp:916]: what about just leaving it as an empty string if no error happens?

File: dbms/src/Storages/Page/V3/PageStorageImpl.cpp [hunk 4]
```diff
@@ -47,6 +47,8 @@ void PageStorageImpl::restore()
 {
     // TODO: clean up blobstore.
     // TODO: Speedup restoring
+    blob_store.registerPaths();
```
Comment (JaySon-Huang) [inline dbms/src/Storages/Page/V3/PageStorageImpl.cpp:50]: If there are BlobFile 1, 2, 3 on disks, but only BlobFile 2, 3 have valid data. Then calling `registerPaths` will register BlobFile 1 on `delegator`, but actually, BlobFile 1 on disk will be removed in `BlobStats::restore`. But it won't be removed from the `delegator`.
https://github.com/pingcap/tiflash/blob/924ea6d67f8931a36523cb1db6cbcbbe4161c740/dbms/src/Storages/Page/V3/PageDirectoryFactory.cpp#L96

Maybe we should call `delegator->addPageFileUsedSize` in `BlobStats::restore`?

Comment (jiaqizho) [reply dbms/src/Storages/Page/V3/PageStorageImpl.cpp:50]: Can't call `addPageFileUsedSize` in `BlobStats::restore`. `addPageFileUsedSize`  or `registerPaths()` can't be call after `restoreByEntry`. Becasue we will find path in `restoreByEntry`.

Added `removePageFile` in `BlobStats::restore()`.

Comment (jiaqizho) [reply dbms/src/Storages/Page/V3/PageStorageImpl.cpp:50]: changed

File: dbms/src/Storages/Page/V3/tests/gtest_blob_store.cpp [hunk 5]
```diff
@@ -67,13 +67,14 @@ try
...
+    BlobFileId file_id2 = 2;
 
-    const auto & path = getTemporaryPath();
-    createIfNotExist(path);
-    Poco::File(fmt::format("{}/{}{}", path, BlobFile::BLOB_PREFIX_NAME, file_id1)).createFile();
-    Poco::File(fmt::format("{}/{}{}", path, BlobFile::BLOB_PREFIX_NAME, file_id2)).createFile();
+    {
+        const auto & lock = stats.lock();
+        stats.createStat(file_id1, lock);
+        stats.createStat(file_id2, lock);
```
Comment (JaySon-Huang) [inline dbms/src/Storages/Page/V3/tests/gtest_blob_store.cpp:76]: maybe we should use `createStatNotChecking` here, and don't change the `file_id1` and `file_id2`?

File: dbms/src/Storages/Page/V3/tests/gtest_page_directory.cpp [hunk 6]
```diff
@@ -1999,19 +1999,24 @@ try
         auto path = getTemporaryPath();
         PSDiskDelegatorPtr delegator = std::make_shared<DB::tests::MockDiskDelegatorSingle>(path);
         BlobStore::BlobStats stats(log, delegator, BlobStore::Config{});
+        {
+            const auto & lock = stats.lock();
+            stats.createStat(file_id1, lock);
+            stats.createStat(file_id2, lock);
```
Comment (JaySon-Huang) [inline dbms/src/Storages/Page/V3/tests/gtest_page_directory.cpp:2005]: ditto

File: dbms/src/Storages/Page/V3/tests/gtest_blob_store.cpp [hunk 7]
```diff
@@ -104,7 +105,6 @@ try
 
     ASSERT_EQ(stats_copy.size(), 1);
     ASSERT_EQ(stats_copy.begin()->second.size(), 2);
-    EXPECT_EQ(stats.roll_id, 13);
```
Comment (JaySon-Huang) [inline dbms/src/Storages/Page/V3/tests/gtest_blob_store.cpp:107]: Add back?

File: dbms/src/Storages/Page/V3/tests/gtest_page_directory.cpp [hunk 8]
```diff
@@ -1960,27 +1960,27 @@ TEST_F(PageDirectoryGCTest, RestoreWithRef)
 try
 {
     BlobFileId file_id1 = 1;
-    BlobFileId file_id2 = 5;
+    BlobFileId file_id2 = 2;
```
Comment (JaySon-Huang) [inline dbms/src/Storages/Page/V3/tests/gtest_page_directory.cpp:1963]: revert back these file id changes?

File: dbms/src/Storages/Page/V3/tests/gtest_page_directory.cpp [hunk 9]
```diff
@@ -1960,27 +1960,27 @@ TEST_F(PageDirectoryGCTest, RestoreWithRef)
...
-    PageEntryV3 entry_5_v1{.file_id = file_id2, .size = 255, .tag = 0, .offset = 0x100, .checksum = 0x4567};
-    PageEntryV3 entry_5_v2{.file_id = file_id2, .size = 255, .tag = 0, .offset = 0x400, .checksum = 0x4567};
+    PageEntryV3 entry_2_v1{.file_id = file_id2, .size = 255, .tag = 0, .offset = 0x100, .checksum = 0x4567};
+    PageEntryV3 entry_2_v2{.file_id = file_id2, .size = 255, .tag = 0, .offset = 0x400, .checksum = 0x4567};
     {
         PageEntriesEdit edit;
-        edit.put(file_id1, entry_1_v1);
-        edit.put(file_id2, entry_5_v1);
+        edit.put(1, entry_1_v1);
+        edit.put(5, entry_2_v1);
```
Comment (JaySon-Huang) [inline dbms/src/Storages/Page/V3/tests/gtest_page_directory.cpp:1976]: revert back these file id changes?

Comment (jiaqizho) [reply dbms/src/Storages/Page/V3/tests/gtest_page_directory.cpp]: revert the `entry_2_v1 ` to `entry_5_v1 `, but need put `page_id` not the `file_id` here , so i changed `file_id` to number

File: dbms/src/Storages/Page/V3/BlobStore.cpp [hunk 10]
```diff
@@ -316,9 +340,13 @@ std::pair<BlobFileId, BlobFileOffset> BlobStore::getPosFromStats(size_t size)
...
-    stat->removePosFromStat(offset, size, lock);
+    {
+        auto lock = stat->lock();
+        stat->removePosFromStat(offset, size, lock);
+    }
 
+    // We don't need hold the BlobStat lock(Also can't do that).
+    // Because once BlobStat become Read-Only type, Then valid size won't increase.
     if (stat->isReadOnly() && stat->sm_valid_size == 0)
     {
```
Comment (JaySon-Huang) [inline dbms/src/Storages/Page/V3/BlobStore.cpp:351]: Maybe the best way to do this is sth like 
```
bool need_remove_stat = false;
{
  auto lock = stat->lock();
  need_remove_stat = stat->removePosFromStat(offset, size, lock);
}
if (need_remove_stat)
{
  // lock `stats` and remove
}
```

**Review by JaySon-Huang — APPROVED** [review]
LGTM

**Pull request merged**